Library/Homebrew/utils.rb

```diff
@@ -404,6 +404,30 @@ def gzip(*paths)
     end
   end
 
+  def reproducible_gzip(*paths)
+    # Ideally, we would just set mtime = 0 if SOURCE_DATE_EPOCH is absent, but Ruby's
+    # Zlib::GzipWriter does not properly handle the case of setting mtime = 0:
+    # https://bugs.ruby-lang.org/issues/16285
+    #
+    # This was fixed in https://github.com/ruby/zlib/pull/10. Set mtime to 0 instead
+    # of raising exception once we are using zlib gem version 1.1.0 or newer.
+    no_env_var_message = "Can't create reproducible gzip file without SOURCE_DATE_EPOCH environment variable set"
+    raise no_env_var_message if ENV["SOURCE_DATE_EPOCH"].blank?
+
+    paths.map do |path|
+      gz = Zlib::GzipWriter.open("#{path}.gz")
+      gz.mtime = ENV["SOURCE_DATE_EPOCH"].to_i
+      gz.orig_name = File.basename(path)
+
+      File.open(path, "rb") do |fp|
+        gz.write(fp.read(64 * 1024)) until fp.eof?
```

Comment:
Would be good to add a comment explaining why these magic numbers are being used and pull them into a variable outside the loop.